File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        if(len(text2) > len(text1)):
            return self.longestCommonSubsequence(text2, text1)
        dp = [[0 for i in range(len(text2)+1)] for j in range(len(text1)+1)]
        ans = 0
        for i,c in enumerate(text1):
            for j,d in enumerate(text2):
                if(c == d):
                    dp[i][j] = 1 + dp[i-1][j-1]
                    ans = max(ans, dp[i][j])
                else:
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1])
                    ans = max(ans, dp[i][j])
        return ans

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    int solve(int i, int j, String a, String b, int[][] dp) {
        int ans = 0;
        if(a.length() == i || b.length() == j)
            return 0;
        
        else if(dp[i][j] != -1)
            return dp[i][j];
        
        else if(a.charAt(i) == b.charAt(j)) {
            dp[i][j] = 1 + solve(i+1, j+1, a, b, dp);
            ans = Math.max(ans, dp[i][j]);
        }
        
        else {
            dp[i][j] = Math.max(solve(i+1, j, a, b, dp), solve(i, j+1, a, b, dp));
            ans = Math.max(ans, dp[i][j]);
        }
        
        return ans;
    }
    public int longestCommonSubsequence(String text1, String text2) {
        int[][] dp = new int[text1.length()][text2.length()];
        for(int[] i:dp) {
            Arrays.fill(i, -1);
        }
        return solve(0, 0, text1, text2, dp);
    }
}